kroxylicious-test-tools/src/main/java/io/kroxylicious/test/ApiMessageSampleGenerator.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.test;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.Message;
+import org.apache.kafka.common.record.BaseRecords;
+import org.apache.kafka.common.record.UnalignedMemoryRecords;
+import org.apache.kafka.common.utils.ImplicitLinkedHashCollection;
+
+/**
+ * Generates an ApiMessage instance per ApiKey. The message is built
+ * with reflection and populated with randomised data. The data is built
+ * from the same random seed and the ApiKeys and methods are iterated over
+ * in a deterministic order so the output of `createRequestSamples` should
+ * be consistent.
+ */
+public class ApiMessageSampleGenerator {
+
+    private ApiMessageSampleGenerator() {
+
+    }
+
+    /**
+     * Generates a sample request ApiMessage for all ApiKeys
+     * @return ApiKeys to message
+     */
+    static public Map<ApiKeys, ApiMessage> createRequestSamples() {
+        Random random = new Random(0);
+        return instantiateAll(DataClasses.getRequestClasses(), random);
+    }
+
+    /**
+     * Generates a sample response ApiMessage for all ApiKeys
+     * @return ApiKeys to message
+     */
+    static public Map<ApiKeys, ApiMessage> createResponseSamples() {
+        Random random = new Random(0);
+        return instantiateAll(DataClasses.getResponseClasses(), random);
+    }
+
+    private static Map<ApiKeys, ApiMessage> instantiateAll(Map<ApiKeys, Class<? extends ApiMessage>> messages, Random random) {
+        return messages.entrySet().stream().sorted(Comparator.comparing(apiKeysClassEntry -> apiKeysClassEntry.getKey().name))
+                .map(entry -> instantiateSample(entry, random)).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    static private Map.Entry<ApiKeys, ApiMessage> instantiateSample(Map.Entry<ApiKeys, Class<? extends ApiMessage>> entry, Random random) {
+        ApiMessage message = (ApiMessage) instantiate(entry.getValue(), random);
+        return new AbstractMap.SimpleEntry<>(entry.getKey(), message);
+    }
+
+    static Set<String> exceptions = new HashSet<>();
```

Comment:
Should add a comment to this. All these exceptional cases cause ApiMessage construction to change because you aren't allowed to set these to non-default values at the latest version for that api.